design-pattern.md

```diff
@@ -1 +1,151 @@
+<!-- _class: due -->
+
 # Design pattern
+
+I design pattern sono dei modelli di soluzione a problemi comuni di progettazione software. Esistono diversi tipi di design pattern, ma in generale si possono suddividere in 4 categorie: _creazionali, strutturali, comportamentali e architetturali_.
+
+## Pattern creazionali
+
+I design pattern creazionali sono quelli che si occupano di risolvere problemi relativi alla _creazione di oggetti_.
+
+### Factory Method
+
+Il **Factory Method** permette di creare oggetti senza specificare la loro classe concreta. Questo pattern è particolarmente utile quando si deve creare un oggetto che dipende da altri oggetti, o quando si deve creare un oggetto che dipende da parametri variabili. Infatti, il Factory Method permette di creare oggetti in modo indipendente dalle dipendenze che essi hanno.
```

Comment:
La prima frase è perfetta (_Il **Factory Method** permette di creare oggetti senza specificare la loro classe concreta._).
Secondo me è un po' poco chiara la parte dopo, tipo _oggetto che dipende da altri oggetti_ o _creare oggetti in modo indipendente dalle dipendenze che essi hanno_.
Cambierei con:

 Questo pattern è particolarmente utile quando abbiamo bisogno di istanziare uno o più oggetti astratti per poi effettuarvi delle operazioni che non dipendono dal loro tipo concreto (uguale per tutti gli oggetti). Come è ben noto non è possibile costruire un oggetto astratto. Per ovviare a questo problema si usa il **Factory Method** che permette al Client di specificare al metodo che dovrà effettuare tali operazioni, solo la _factory_ con cui costruire gli oggetti astratti, il metodo poi potrà lavorare con gli oggetti astratti e non sarà necessario creare un metodo diverso (che effettua le stesse operazioni) per ciascun tipo concreto, solo per avere accesso al costruttore del tipo concreto in questione.